controllers/trivy_vuln_crds_controller.go

```diff
@@ -0,0 +1,219 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/aquasecurity/trivy-operator/pkg/apis/aquasecurity/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"os"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strconv"
+)
+
+const LagoonProjectLabel = "lagoon.sh/project"
+const TrivyVulnReportNamespace = "trivy-operator.resource.namespace"
+
+type TrivyVulnerabilityReportReconciler struct {
+	Client         client.Client
+	Scheme         *runtime.Scheme
+	MessageQWriter func(data []byte) error
+	WriteToQueue   bool
+	PredicateData  interface{}
+}
+
+type ProblemsPayload struct {
+	Problems []interface{} `json:"problems"`
+}
+
+// predicate that filters out non-lagoon projects
+func (r *TrivyVulnerabilityReportReconciler) lagoonProjectLabelsOnlyPredicate() predicate.Predicate {
+
+	return predicate.Funcs{
+		CreateFunc: func(event event.CreateEvent) bool {
+			if r.isLagoonProject(event.Object) &&
+				!insightsProcessedAnnotationExists(event.Object) {
+				return true
+			}
+			return false
+		},
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			if r.isLagoonProject(event.ObjectNew) &&
+				!insightsProcessedAnnotationExists(event.ObjectNew) {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+func (r *TrivyVulnerabilityReportReconciler) isLagoonProject(event client.Object) bool {
+	var namespace string
+	for k, v := range event.GetLabels() {
+		if k == TrivyVulnReportNamespace {
+			namespace = v
+		}
+	}
+
+	nsObj := &corev1.Namespace{}
+	nsObj.SetName(namespace)
+
+	// retrieve the ns object from the k8 API
+	err := r.Client.Get(context.Background(), client.ObjectKeyFromObject(nsObj), nsObj)
+	if err != nil {
+		fmt.Printf("Failed to get Namespace %q: %v\n", namespace, err)
+		os.Exit(1)
+	}
+
+	// check the ns if this is a lagoon project
+	isLagoonProject := false
+	for k, v := range nsObj.Labels {
+		if k == LagoonProjectLabel && v == namespace {
+			isLagoonProject = true
+		}
+	}
+
+	if isLagoonProject {
+		// store some data since we've already done the lookup, and it doesn't exist in the vuln. report
+		r.PredicateData = nsObj.Labels
```

Comment:
So I think this may be a real problem that just struck me. Although this is really only a guess (I haven't tested).

Here's we're saving data against the reconciler object _itself_ - the problem is, though, that this reconciler object (I believe) deals with all incoming reconciliations.
I'm guessing that we're potentially going to see clashes here - at any point, any reconciliation can be writing to this field.

Reply:
(we can chat about this when you're back)

Reply:
Yeah good spot - we could make the PredicateData a sync.Map data struct instead of an interface{} to store the Labels. Then each PredicateData can be stored separately for each reconciliation. So something like:
```

 type TrivyVulnerabilityReportReconciler struct {
-       Client         client.Client
-       Scheme         *runtime.Scheme
-       MessageQWriter func(data []byte) error
-       WriteToQueue   bool
-       PredicateData  interface{}
+       Client           client.Client
+       Scheme           *runtime.Scheme
+       MessageQWriter   func(data []byte) error
+       WriteToQueue     bool
+       PredicateDataMap sync.Map
 }

 type ProblemsPayload struct {
@@ -86,8 +87,7 @@ func (r *TrivyVulnerabilityReportReconciler) isLagoonProject(event client.Object

        if isLagoonProject {
                // store some data since we've already done the lookup, and it doesn't exist in the vuln. report
-               r.PredicateData = nsObj.Labels
-
+               r.PredicateDataMap.Store(client.ObjectKeyFromObject(event), nsObj.Labels)
                return true
        }

@@ -115,7 +115,7 @@ func (r *TrivyVulnerabilityReportReconciler) Reconcile(ctx context.Context, requ
                                return reconcile.Result{}, err
 ```

Reply:
But also, from looking at the other comment on moving the isLagoonProject method to a generic function, the above will also need adjusting